### PR TITLE
Separated templates and directory list

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Closes https://github.com/enso-org/cloud-v2/issues/490:
- Make templates collapsible
- Add scroll-based shadow to templates
- Move scrollbars from body to templates and directory listing

### Important Notes
None

### Screenshots/Screencasts

[screen-recorder-thu-jun-08-2023-19-46-49.webm](https://github.com/enso-org/enso/assets/4046547/f9b548f5-6c16-4e93-bda5-e1f3d705a69a)

"New blank project" button:
image

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

**Review by PabloBuchu — APPROVED** [review]

**Comment by somebody1234** [comment]
@indiv0 could you QA this?

**Comment by jdunkerley** [comment]
Could we have an "add new empty project" button in collapsed mode? Like in Excel:
image

And will this state be remembered between sessions?

**Comment by somebody1234** [comment]
good idea - and yes, the state is already remembered between sessions (using `localStorage` as usual)

**Comment by indiv0** [comment]
**TL;DR**:  ❌ QA Failed

# Questions/Notes

- Should the `Drive` section have the same scroll shadows as `Templates`? It currently doesn't.
- ~~Will @jdunkerly's suggestion to add a "New blank workbook" button be added in this PR or a future PR? If it's the latter, a tracking issue should be opened.~~
    - Now being tracked as https://github.com/enso-org/cloud-v2/issues/507
- ~~Collapsing the "Templates" in the local view makes them collapsed in the cloud view. Not sure if this is intentional. Seems a bit weird since I consider the two to be separate pages entirely, but I guess some users might expect this behaviour.~~
    - Per https://github.com/enso-org/enso/pull/6995#issuecomment-1595456616, this is working as intended

# Scenario Descriptions

<table style="border: 1px solid black;">
    <thead>
        <tr>
            <th style="border: 1px solid black;" rowspan=4>Scenario</th>
        </tr>
        <tr>
            <th style="border: 1px solid black;" colspan=5>Command</th>
            <th style="border: 1px solid black;" colspan=2>Features</th>
            <th style="border: 1px solid black;" rowspan=3>Debug Scene</th>
        </tr>
        <tr style="border: 1px solid black;">
            <th style="border: 1px solid black;">GUI</th>
            <th style="border: 1px solid black;" colspan=2>IDE</th>
            <th style="border: 1px solid black;" colspan=2>NPM</th>
            <th style="border: 1px solid black;" rowspan=2>Auth</th>
            <th style="border: 1px solid black;" rowspan=2>New Dashboard</th>
        </tr>
        <tr style="border: 1px solid black;">
            <th style="border: 1px solid black;">Watch</th>
            <th style="border: 1px solid black;">Watch</th>
            <th style="border: 1px solid black;">Build</th>
            <th style="border: 1px solid black;">build-dashboard</th>
            <th style="border: 1px solid black;">watch-dashboard</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td style="border: 1px solid black;">0</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">❌</td>
            <td style="border: 1px solid black;">❌</td>
            <td style="border: 1px solid black;">❌</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">1</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">2</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">❌</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">3</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">❌</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">4</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">5</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
        </tr>
    </tbody>
</table>

# Scenario 0

```sh
./run gui watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
```

- ✅ GUI at http://localhost:8080 opens
- ✅ Can login
- ✅ Dashboard shows up
- ✅ Templates are collapsible
- ✅ Templates have scroll-based shadow
- ✅ Template collapse state persists across reloads
- ✅ Drive is scrollable
- ✅ Body is not scrollable

<details>
<summary>
Screenshot
</summary>
image
</details>

# Scenario 1

```sh
ENSO_BUILD_TARGET=dir \
    ./run ide watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --backend-source=release \
    --backend-release=nightly \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info \
    --ide-option=-feature-preview.new-dashboard \
    --ide-option=-debug.dev-tools \
    --ide-option=-debug \
    --ide-option=--inspect=5858 \
    --ide-option=-authentication \
    --ide-option=-startup.entry=text_area
```

- ✅ Debug scene is visible
- ✅ Logs appear with `console.log("Hello")`

<details>
<summary>
Screenshot
</summary>
image

</details>

# Scenario 2

```sh
ENSO_BUILD_TARGET=dir \
    ./run ide watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --backend-source=release \
    --backend-release=nightly \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info \
    --ide-option=-feature-preview.new-dashboard \
    --ide-option=-debug.dev-tools \
    --ide-option=-debug \
    --ide-option=--inspect=5858 \
    --ide-option=-authentication
```

- ✅ Templates are collapsible
- ✅ Templates have scroll-based shadow
- ✅ Template collapse state persists across reloads
- ✅ Drive is scrollable
- ✅ Body is not scrollable
- ✅ Nodes in Cloud project are visible

<details>
<summary>
Screenshot
</summary>
https://github.com/enso-org/enso/assets/641547/f671635f-b51e-405c-a226-317409ff62ae
</details>

# Scenario 3

```sh
ENSO_BUILD_TARGET=dir \
    ./run ide build \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --backend-source=release \
    --backend-release=nightly \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
./dist/ide/mac-arm64/Enso.app/Contents/MacOS/Enso \
    -feature-preview.new-dashboard \
    -debug.dev-tools \
    -debug \
    --inspect=5858 \
    -authentication
```

- ✅ Templates are collapsible
- ✅ Templates have scroll-based shadow
- ✅ Template collapse state persists across reloads
- ✅ Drive is scrollable
- ✅ Body is not scrollable
- ✅ Nodes in Cloud project are visible

<details>
<summary>
Screenshot
</summary>
image
</details>

# Scenario 4

```sh
cd app/ide-desktop
npm run build-dashboard
```

- ✅ Command ran successfully

<details>
<summary>
Screenshot
</summary>
image

</details>

# Scenario 5

```sh
cd app/ide-desktop
npm run watch-dashboard
```

- ✅ Command ran successfully
- ✅ GUI at http://localhost:8080 opens
- ✅ Templates are collapsible
- ✅ Templates have scroll-based shadow
- ✅ Template collapse state persists across reloads
- ✅ Drive is scrollable
- ✅ Body is not scrollable
- ❌ Nodes in Cloud project are visible
  - Failed to open Cloud project due to failure to open `.glsl` files -- a service worker aborted the operation. See screenshot for details.

<details>
<summary>
Screenshot
</summary>
image
</details>

**Comment by somebody1234** [comment]
drive section having scroll shadows: the figma design didn't have them so i didn't add them - @PabloBuchu do you think it should be added?
new blank workbook: yeah it should probably be here. will add sometime
templates collapsed state is the same in both local and cloud view: intentional - while your point is valid, i feel like if they don't want to use templates for local dev they probably don't want to use templates for remote dev?
glsl error: will try reproing sometime

**Comment by indiv0** [comment]
new blank workbook: Could we open an issue to track this?

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx [hunk 1]
```diff
@@ -1097,334 +1097,357 @@ function Dashboard(props: DashboardProps) {
...
-                                onContextMenu={(projectAsset, event) => {
-                                    event.preventDefault()
-                                    event.stopPropagation()
-                                    const doOpenForEditing = () => {
-                                        unsetModal()
-                                        setProjectEvent({
-                                            type: projectActionButton.ProjectEventType.open,
-                                            projectId: projectAsset.id,
-                                        })
+                    <div className="h-6 mx-2" />
```
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx:1228]: Empty div, is this for padding purposes?

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx [hunk 2]
```diff
@@ -1097,334 +1097,357 @@ function Dashboard(props: DashboardProps) {
...
+                                        <td key={column} className={COLUMN_CSS_CLASS[column]} />
+                                    ))}
+                                </tr>
+                                <Rows<backendModule.Asset<backendModule.AssetType.project>>
+                                    items={visibleProjectAssets}
+                                    getKey={projectAsset => projectAsset.id}
+                                    isLoading={isLoadingAssets}
+                                    placeholder={
+                                        <span className="opacity-75">
+                                            You have no project yet. Go ahead and create one using
```
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx]: `project` -> `projects` maybe?

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx [hunk 3]
```diff
@@ -1097,334 +1097,357 @@ function Dashboard(props: DashboardProps) {
...
+                                        event.stopPropagation()
+                                        const doOpenForEditing = () => {
+                                            unsetModal()
+                                            setProjectEvent({
+                                                type: projectActionButton.ProjectEventType.open,
+                                                projectId: projectAsset.id,
+                                            })
+                                        }
+                                        const doOpenAsFolder = () => {
+                                            // FIXME[sb]: Uncomment once backend support
```
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx]: FIXMEs should be fixed before a PR is merged.
TODOs can be left in if they're being addressed in a later PR.
TODOs must be associated with an issue in the issue tracker so we can keep track of them.

Seems like this should be changed to a TODO and associated with an issue. This applies to the other FIXMEs/TODOs touched by this PR.

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/templates.tsx [hunk 4]
```diff
@@ -1,6 +1,30 @@
...
+import * as common from 'enso-common'
+
 import * as svg from '../../components/svg'
 
+// =================
+// === Constants ===
+// =================
+
+/** The `localStorage` key used to store whether the {@link Templates} element should be
+ * not collapsed by default. */
```
Comment (indiv0) [inline app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/templates.tsx]: `should be not collapsed` -> `should be expanded`

**Review by indiv0 — CHANGES_REQUESTED** [review]

File: app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx (hunk 1 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx:1228]: yes, comment added

**Comment by somebody1234** [comment]
> Failed to open Cloud project due to failure to open `.glsl` files
@indiv0 Unfortunately I can't repro - you might have to do some further investigation? Or it might just be that that specific version is missing some shaders in the S3 bucket

I do get different errors on FF though:
image

But given that they come from Rust I don't think this PR is the cause

**Comment by indiv0** [comment]
Attempted a re-QA of scenario 5 to reproduce the `.glsl` issue. Issue did not reoccur but I ran into the same selection menu issues reported in #7060, so proper QA is blocked until that issue is resolved. Here is a screenshot with error logs for reference:

image

**Comment by PabloBuchu** [comment]
I have 15' mac and the templates looks like this.. (the 10px scroll looks nasty imho).. also there is no margin / padding around... @somebody1234 even on your first screen recording I see such small scroll... can we somehow make it to fit the screen and make the scroll appear if the user display is smaller?

https://github.com/enso-org/enso/assets/2855109/4daeb1f6-0f88-4b87-9a28-4a5f930bbbd4

**Comment by PabloBuchu** [comment]
@indiv0 @somebody1234 regarding the scenario 5 I was able to open cloud project and do stuff so I think the bug was fixed.

waiting for the fix for templates and then we could merge

Screenshot 2023-06-26 at 12 47 59

**Comment by wdanilo** [comment]
When templates are collapsed, I don't think we need to create a separate button for new project, as there is already `+` button next to the "projects" header.

**Review by wdanilo — APPROVED** [review]
Also, the video does not work in Safari, it works in Chrome. How did you record / store it? We should fix that in the next PRs so videos will be viewable on all browsers. I know that Safari is the pain in the ass here :(

**Comment by somebody1234** [comment]
i use the "screen recorder" chrome extension, but the issue is probably that it's stored as a webm. it's easy enough for me to convert to mp4 though

**Comment by PabloBuchu** [comment]
@somebody1234  Why suddenly templates are narrowed? Is this issue only on my machine? Also as Wojtek said please remove this `New blank project` button

Screenshot 2023-06-30 at 11 56 32

**Comment by somebody1234** [comment]
@PabloBuchu first issue: was on `develop` but not here. fixed after merge
second issue: "new project" button removed.

**Comment by PabloBuchu** [comment]
QA 🟢 @somebody1234 please merge once the CI is completed

**Review by indiv0 — APPROVED** [review]
